REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [52]:
import pandas as pd

In [53]:
bus = pd.read_csv('businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('inspections.csv')

In [54]:
bus.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [55]:
ins.head()

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [56]:
bus.shape[0]

6315

#### Question 2b: How many unique business IDs are there?  

In [57]:
bus['business_id'].nunique()

6315

#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [58]:
bus['name'].value_counts().head(5)

STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
Name: name, dtype: int64

In [59]:
bus.columns

Index(['business_id', 'name', 'address', 'city', 'state', 'postal_code',
       'latitude', 'longitude', 'phone_number'],
      dtype='object')

In [60]:
bus[bus.city=='San Francisco']['name'].value_counts().head(5)

STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
Name: name, dtype: int64

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [61]:
type(bus.postal_code[0])

str

#### Question 3b

What are the unique values of postal_code?

In [62]:
set(bus.postal_code)

{'00000',
 '92672',
 '94013',
 '94014',
 '94066',
 '941',
 '94101',
 '94102',
 '94103',
 '941033148',
 '94104',
 '94105',
 '94107',
 '94108',
 '94109',
 '94110',
 '941102019',
 '94111',
 '94112',
 '94114',
 '94115',
 '94116',
 '94117',
 '94118',
 '94120',
 '94121',
 '94122',
 '94123',
 '94124',
 '94127',
 '94129',
 '94130',
 '94131',
 '94132',
 '94133',
 '94134',
 '94143',
 '94158',
 '94188',
 '94545',
 '94602',
 '94609',
 '94621',
 '95105',
 'CA',
 'Ca',
 nan}

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [63]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [64]:
bus_valid=bus[bus['postal_code'].isin(validZip)]

In [65]:
bus_valid

business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              31      NORMAN'S ICE CREAM AND FREEZES   
4              45                 CHARLIE'S DELI CAFE   
...           ...                                 ...   
6310        90234                ACCESS SFUSD THE ARC   
6311        90236                      PHO DONG HUONG   
6312        90237                    WIN'S RESTAURANT   
6313        90238                    USA POWER MARKET   
6314        90239                          ALBA RAY'S   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3              2801 LEAVENWORTH ST   San Francisco    CA       94133   
4                   3202 FOLSOM ST   San Francisco    CA       94110   
...                             ...            ...   ...         ...   
6310                1520 HOWARD ST   San Francisco    CA       94103   
6311             667 MONTEREY BLVD   San Francisco    CA       94127   
6312               3040 TARAVAL ST   San Francisco    CA       94116   
6313                1524 OCEAN AVE   San Francisco    CA       94112   
6314               2293 MISSION ST   San Francisco    CA       94110   

       latitude   longitude  phone_number  
0     37.791116 -122.403816  +14154217044  
1     37.786848 -122.421547  +14157763262  
2     37.792888 -122.403135  +14156779494  
3     37.807155 -122.419004           NaN  
4     37.747114 -122.413641  +14156415051  
...         ...         ...           ...  
6310        NaN         NaN           NaN  
6311        NaN         NaN  +14154945048  
6312        NaN         NaN  +14158606036  
6313        NaN         NaN  +14153331878  
6314        NaN         NaN  +14158881393  

[5999 rows x 9 columns]

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [69]:
df_nl = bus_valid[pd.isnull(bus_valid['longitude'])]
df_l = bus_valid[~pd.isnull(bus_valid['longitude'])]

In [70]:
df_nl

business_id                               name             address  \
200           821                       LA SANTANECA    3781 MISSION ST    
210           852  NEW HOLLYWOOD BAKERY & RESTAURANT    652 PACIFIC AVE    
760          2934                   LOU'S FISH SHACK   300 JEFFERSON ST    
826          3123                        LES JOULINS        44 ELLIS ST    
1100         4323                             DIVA'S       1081 POST ST    
...           ...                                ...                 ...   
6310        90234               ACCESS SFUSD THE ARC     1520 HOWARD ST    
6311        90236                     PHO DONG HUONG  667 MONTEREY BLVD    
6312        90237                   WIN'S RESTAURANT    3040 TARAVAL ST    
6313        90238                   USA POWER MARKET     1524 OCEAN AVE    
6314        90239                         ALBA RAY'S    2293 MISSION ST    

               city state postal_code  latitude  longitude  phone_number  
200   San Francisco    CA       94110       NaN        NaN  +14156481034  
210   San Francisco    CA       94133       NaN        NaN  +14153979919  
760   San Francisco    CA       94133       NaN        NaN  +14157715687  
826   San Francisco    CA       94102       NaN        NaN  +14153975397  
1100  San Francisco    CA       94109       NaN        NaN  +14154743482  
...             ...   ...         ...       ...        ...           ...  
6310  San Francisco    CA       94103       NaN        NaN           NaN  
6311  San Francisco    CA       94127       NaN        NaN  +14154945048  
6312  San Francisco    CA       94116       NaN        NaN  +14158606036  
6313  San Francisco    CA       94112       NaN        NaN  +14153331878  
6314  San Francisco    CA       94110       NaN        NaN  +14158881393  

[2483 rows x 9 columns]

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [74]:
num_nl = df_nl.groupby(['postal_code']).business_id.count()
num_l = df_l.groupby(['postal_code']).business_id.count()

In [83]:
num_nl.index==num_l.index

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True])

In [84]:
newdf=pd.DataFrame({'postal_code':validZip,'null_lon':list(num_nl),'not_null_lon':list(num_l)})

In [85]:
newdf

postal_code  null_lon  not_null_lon
0        94102       196           262
1        94103       252           306
2        94104        59            74
3        94105       100           126
4        94107       247           204
5        94108        82           142
6        94109       146           237
7        94110       251           334
8        94111       115           167
9        94112        59           131
10       94114       108           122
11       94115        76           162
12       94116        35            58
13       94117        67           130
14       94118       100           143
15       94121        55           106
16       94122       123           147
17       94123        53           116
18       94124        88            76
19       94127        28            43
20       94131        14            34
21       94132        65            69
22       94133       135           282
23       94134        29            45

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [87]:
import numpy as np

In [90]:
prop= np.array(newdf.null_lon)/(np.array(newdf.not_null_lon)+np.array(newdf.null_lon))

In [99]:
order = np.flip(np.argsort(prop))
print(prop[order][:5])
print(np.array(newdf.postal_code)[order][:5])

[0.54767184 0.53658537 0.48507463 0.46956522 0.45555556]
['94107' '94124' '94132' '94114' '94122']


# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [105]:
ins.shape[0]==ins['business_id'].nunique()

False

#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [108]:
set(ins['type'])

{'complaint', 'routine'}

In [111]:
ins[ins.type=='complaint'].shape[0]

1

In [112]:
ins[ins.type=='routine'].shape[0]

15429

In [113]:
ins2=ins[ins.type=='routine']

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [119]:
ins['year']=[int(str(dt)[:4]) for dt in ins['date']]

In [120]:
ins

business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15430 rows x 5 columns]

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [121]:
set(ins['year'])

{2013, 2014, 2015, 2016}

In [123]:
for y in set(ins['year']):
    print(y,':',ins[ins.year==y].shape[0])
        

2016 : 5840
2013 : 38
2014 : 5629
2015 : 3923


In [124]:
ins3=ins[ins.year!=2013]

In [126]:
ins3

business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15392 rows x 5 columns]

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [140]:
ins2016=ins[ins.year==2016]

In [141]:
ins2016.groupby(['business_id']).count().score.describe()

count    4543.000000
mean        1.285494
std         0.490489
min         1.000000
25%         1.000000
50%         1.000000
75%         2.000000
max         3.000000
Name: score, dtype: float64

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [147]:
merged = pd.merge(bus, ins2016, on='business_id')

merged_head = merged.head(8)

merged_head

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4           31      NORMAN'S ICE CREAM AND FREEZES   
5           45                 CHARLIE'S DELI CAFE   
6           45                 CHARLIE'S DELI CAFE   
7           48                          ART'S CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
4           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
5                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   
6                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   
7                 747 IRVING ST   San Francisco    CA       94122  37.764013   

    longitude  phone_number  score      date     type  year  
0 -122.403816  +14154217044     82  20160503  routine  2016  
1 -122.421547  +14157763262     94  20160513  routine  2016  
2 -122.403135  +14156779494     98  20161005  routine  2016  
3 -122.403135  +14156779494     96  20160311  routine  2016  
4 -122.419004           NaN     84  20160816  routine  2016  
5 -122.413641  +14156415051     84  20160614  routine  2016  
6 -122.413641  +14156415051     78  20160104  routine  2016  
7 -122.465749  +14156657440     94  20160630  routine  2016

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [150]:
order = np.argsort(merged.score)
merged_st = merged.loc[order,['name','address','score']]

In [152]:
merged_st.head(20)

name              address  score
3931      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD      52
5727                    POKI TIME     2101 LOMBARD ST      55
4981                 CRAZY PEPPER   2257 SAN JOSE AVE      55
3917                   UNCLE CAFE       65 WAVERLY PL      55
2488             HING WANG BAKERY        339 JUDAH ST      56
4481                   GOLDEN WOK   295 B ORIZABA AVE      56
4298                   TOP'S CAFE  2460 SAN BRUNO AVE      57
4824    L & G VIETNAMESE SANDWICH         602 EDDY ST      57
696                  RED A BAKERY      634 CLEMENT ST      57
3415      SAN TUNG RESTAURANT LLC      1031 IRVING ST      57
457           MA'S DIM SUM & CAFE      1315 POWELL ST      57
3940  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST      57
5226    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST      58
523           JOANIE'S DINER INC.   1329 COLUMBUS AVE      59
5424         PAK NAM THAI CUISINE       655 LARKIN ST      59
2927                  JADE GARDEN      3608 BALBOA ST      59
3430                        ITHAI         720 POST ST      59
4885          VALLARTA'S TACO BAR     2022 MISSION ST      60
3904             HONG KONG LOUNGE     5322 GEARY BLVD      60
4025                 WOLFES LUNCH        1220 16TH ST      60

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!

In [ ]:
commit